packages/backend/src/models/user.ts

```diff
@@ -148,40 +148,136 @@ class User extends Base {
     await super.$beforeUpdate(opt, queryContext)
   }
 
-  withAccessibleFlow({
+  /**
+   * we use this filter to check for two things:
+   * 1. user is a valid owner or collaborator on this pipe
+   * 2. user has the required permissions to work on this pipe
+   */
+  private applyAccessibilityFilter(
+    query:
+      | ExtendedQueryBuilder<Flow, Flow[]>
+      | ExtendedQueryBuilder<Step, Step[]>
+      | ExtendedQueryBuilder<Connection, Connection[]>,
+    userId: string,
+    requiredRole: IFlowCollabRole,
+  ) {
+    if (requiredRole === 'owner') {
+      query.where('flows.user_id', userId)
+    } else if (requiredRole === 'editor') {
+      query.where(function () {
+        this.where('flows.user_id', userId).orWhereExists(function () {
+          this.select('*')
+            .from('flow_collaborators')
+            .whereRaw('flow_collaborators.flow_id = flows.id')
+            .where('flow_collaborators.user_id', userId)
+            .where('flow_collaborators.role', 'editor')
+            .whereNull('flow_collaborators.deleted_at')
+        })
+      })
+    } else if (requiredRole === 'viewer') {
+      query.where(function () {
+        this.where('flows.user_id', userId).orWhereExists(function () {
+          this.select('*')
+            .from('flow_collaborators')
+            .whereRaw('flow_collaborators.flow_id = flows.id')
+            .where('flow_collaborators.user_id', userId)
+            .whereIn('flow_collaborators.role', ['editor', 'viewer'])
+            .whereNull('flow_collaborators.deleted_at')
+        })
+      })
```

Comment:
this can be simplified.

.where('flows.user_id', userId) is always present

.whereIn('...', ['editor', 'viewer']) <--- only this array changes if not owner perms

probably can see how to re-use the existing `checkUserPermission` helper file by adding a method to getAllowedRoles(requiredRole) => ['viewer', 'editor']